pgxn/neon/pagestore_smgr.c

```diff
@@ -1583,16 +1574,103 @@ neon_get_request_lsn(bool *latest, NRelFileInfo rinfo, ForkNumber forknum, Block
 #else
 		flushlsn = GetFlushRecPtr();
 #endif
-		if (lsn > flushlsn)
+		if (last_written_lsn > flushlsn)
 		{
 			neon_log(DEBUG5, "last-written LSN %X/%X is ahead of last flushed LSN %X/%X",
-				 (uint32) (lsn >> 32), (uint32) lsn,
-				 (uint32) (flushlsn >> 32), (uint32) flushlsn);
-			XLogFlush(lsn);
+					 LSN_FORMAT_ARGS(last_written_lsn),
+					 LSN_FORMAT_ARGS(flushlsn));
+			XLogFlush(last_written_lsn);
+			flushlsn = last_written_lsn;
 		}
+
+		/*
+		 * Request the latest version of the page. The most up-to-date request
+		 * LSN we could use would be the current insert LSN, but to avoid the
+		 * overhead of looking it up, use 'flushlsn' instead. This relies on the
+		 * assumption that if the page was modified since the last WAL flush, it
+		 * should still be in the buffer cache, and we wouldn't be requesting it.
+		 */
+		*request_lsn = flushlsn;
+		*not_modified_since = last_written_lsn;
 	}
+}
 
-	return lsn;
+/*
+ *  neon_prefetch_response_usable -- Can a new request be satisfied by old one?
+ *
+ * This is used to check if the response to a prefetch request can be used to
+ * satisfy a page read now.
+ */
+static bool
+neon_prefetch_response_usable(XLogRecPtr request_lsn, XLogRecPtr not_modified_since,
+							  PrefetchRequest *slot)
+{
+	/* sanity check the LSN's on the old and the new request */
+	Assert(request_lsn >= not_modified_since);
+	Assert(slot->request_lsn >= slot->not_modified_since);
+
+	/*
+	 * The new request's LSN should never be older than the old
+	 * one. Similarly, the not_modified_since value calculated for a page
+	 * should never move backwards. This could be an Assert, except that for
```

Comment:
Hmm, the `not_modified_since` value could certainly move _forwards_ but I don't see how it would ever move backwards. Can you sketch step-by-step how that would happen?

Reply:
We put the page to the LFC when we read the response to a GetPage request. It would be better to do it when the page is evicted from the shared buffer cache, but there is no convenient hook there to do it.

We don't update the last-written LSN cache when a page is read, only in smgrwrite(). And smgrextend and some other places like index builds, where pages are written.

So yeah, I see that assumption is fragile. If we did things slightly differently, it could move backwards. I'll keep it as a LOG message, but we should not throw errors or panic. It should not happen in production AFAICS, but if it does, I'd love to find out about it.